cumulus/pallets/parachain-system/src/validate_block/implementation.rs

```diff
@@ -179,59 +144,184 @@ where
+
+		run_with_externalities_and_recorder::<B, _, _>(
+			&backend,
+			&mut recorder,
+			&mut Default::default(),
+			|| {
+				let relay_chain_proof = crate::RelayChainStateProof::new(
+					PSC::SelfParaId::get(),
+					inherent_data.validation_data.relay_parent_storage_root,
+					inherent_data.relay_chain_state.clone(),
+				)
+				.expect("Invalid relay chain state proof");
+
+				#[allow(deprecated)]
+				let res = CI::check_inherents(&block, &relay_chain_proof);
+
+				if !res.ok() {
+					if log::log_enabled!(log::Level::Error) {
+						res.into_errors().for_each(|e| {
+							log::error!("Checking inherent with identifier `{:?}` failed", e.0)
+						});
+					}
+
+					panic!("Checking inherents failed");
+				}
+			},
 		);
-		let hrmp_watermark = crate::HrmpWatermark::<PSC>::get();
-
-		let head_data =
-			if let Some(custom_head_data) = crate::CustomValidationHeadData::<PSC>::get() {
-				HeadData(custom_head_data)
-			} else {
-				head_data
-			};
-
-		ValidationResult {
-			head_data,
-			new_validation_code: new_validation_code.map(Into::into),
-			upward_messages,
-			processed_downward_messages,
-			horizontal_messages,
-			hrmp_watermark,
-		}
-	})
+
+		run_with_externalities_and_recorder::<B, _, _>(
+			&backend,
+			&mut recorder,
+			&mut overlay,
+			|| {
+				parent_header = block.header().clone();
+
+				E::execute_block(block);
+
+				new_validation_code =
+					new_validation_code.take().or(crate::NewValidationCode::<PSC>::get());
+
+				let mut found_separator = false;
+				crate::UpwardMessages::<PSC>::get()
+					.into_iter()
+					.filter_map(|m| {
+						if cfg!(feature = "experimental-ump-signals") {
+							if m == UMP_SEPARATOR {
+								found_separator = true;
+								None
+							} else if found_separator {
+								if upward_message_signals.iter().all(|s| *s != m) {
+									upward_message_signals.push(m);
+								}
+								None
+							} else {
+								Some(m)
+							}
+						} else {
+							Some(m)
+						}
+					})
+					.for_each(|m| {
+						upward_messages.try_push(m)
+							.expect(
+								"Number of upward messages should not be greater than `MAX_UPWARD_MESSAGE_NUM`",
+							)
+					});
+
+				processed_downward_messages += crate::ProcessedDownwardMessages::<PSC>::get();
+				horizontal_messages.try_extend(crate::HrmpOutboundMessages::<PSC>::get().into_iter()).expect(
+					"Number of horizontal messages should not be greater than `MAX_HORIZONTAL_MESSAGE_NUM`",
+				);
+				hrmp_watermark = crate::HrmpWatermark::<PSC>::get();
+
+				if block_index + 1 == num_blocks {
+					head_data = Some(
+						crate::CustomValidationHeadData::<PSC>::get()
+							.map_or_else(|| HeadData(parent_header.encode()), HeadData),
+					);
+				}
+			},
+		)
+	}
+
+	if !upward_message_signals.is_empty() {
+		let mut selected_core = None;
+
+		upward_message_signals.iter().for_each(|s| {
+			if let Ok(UMPSignal::SelectCore(selector, offset)) = UMPSignal::decode(&mut &s[..]) {
```

Comment:
nit: maybe match would be better here? When UMPSignal will get another variant, this if could be easily omitted. match here would emit compiler error in that case?

Reply:
Hmm. Yeah maybe, actually I only want to match on `core` here, but yeah.